File: scripts/test_dump.mjs
#!/usr/bin/env node
/**
 * Test script for raw card dump via JavaScript SDK.
 *
 * Tests:
 *   1. HTTP: GET /v1/readers/0/dump (fetch)
 *   2. WS: dumpCard() command
 *   3. WS: subscribe({ includeRaw: true }) + card_data event (optional, tap a card)
 *   4. Creality CFS decode from MIFARE Classic raw blocks
 *
 * Usage:
 *   cd /path/to/nfc-agent
 *   npm run build --prefix sdk/javascript   # build SDK first
 *   node scripts/test_dump.mjs
 *   node scripts/test_dump.mjs --test subscribe
 */

import { createRequire } from 'module';

// Load the local SDK build directly
const require = createRequire(import.meta.url);
let NFCAgentWebSocket;

try {
  const sdk = await import('../sdk/javascript/dist/index.js');
  NFCAgentWebSocket = sdk.NFCAgentWebSocket;
} catch (err) {
  console.error('ERROR: Could not load JS SDK. Run: npm run build --prefix sdk/javascript');
  console.error(err.message);
  process.exit(1);
}

// Node.js 18+ has global fetch; polyfill for older versions
if (typeof fetch === 'undefined') {
  console.error('ERROR: fetch not available. Use Node.js 18+');
  process.exit(1);
}

const BASE_URL = 'http://127.0.0.1:32145/v1';

// Creality CFS AES key for data blocks (from CrealityMaterialStandard.php)
const CREALITY_AES_KEY_DATA = Buffer.from('H@CFkRnz@KAtBJp2');

const CREALITY_FILAMENT_NAMES = {
  '010101': 'PLA',
  '010201': 'PETG',
  '010301': 'ABS',
  '010401': 'TPU',
  '010501': 'ASA',
  '010601': 'PC',
  '010701': 'PA',
  '010801': 'PLA-CF',
  '010901': 'PETG-CF',
  '010A01': 'ABS-CF',
  '010B01': 'PA-CF',
  '010C01': 'PLA-Silk',
  '010D01': 'PLA-Matte',
};

const LENGTH_CODE_TO_WEIGHT = {
  '0330': '1000g',
  '0247': '750g',
  '0198': '600g',
  '0165': '500g',
  '0082': '250g',
};

// ============================================================================
// Creality decode
// ============================================================================

function decryptCrealityBlock(hexData) {
  const { createDecipheriv } = require('crypto');
  const enc = Buffer.from(hexData, 'hex');
  const decipher = createDecipheriv('aes-128-ecb', CREALITY_AES_KEY_DATA, null);
  decipher.setAutoPadding(false);
  return Buffer.concat([decipher.update(enc), decipher.final()]).toString('ascii');
}

function tryDecodeCreality(blocks) {
  const blockNums = [4, 5, 6];
  for (const n of blockNums) {
    if (!blocks[n]) {
      console.log(`  Block ${n} not available — Creality decode skipped`);
      return;
    }
  }

  let decrypted = '';
  for (const n of blockNums) {
    const dec = decryptCrealityBlock(blocks[n]);
    console.log(`  Block ${n} decrypted: ${JSON.stringify(dec)}`);
    decrypted += dec;
  }

  console.log(`\n  Full payload (48 chars): ${JSON.stringify(decrypted)}`);

  if (decrypted.length < 34) {
    console.log(`  Payload too short: ${decrypted.length} chars`);
    return;
  }

  const payload = decrypted.padEnd(48, '0').slice(0, 48);
  const filamentId = payload.slice(12, 18).toUpperCase();
  const colorHex = payload.slice(18, 24).toUpperCase();
  const lengthCode = payload.slice(24, 28);

  console.log('\n  Parsed Creality fields:');
  console.log(`    Date Code:     ${payload.slice(0, 6)}`);
  console.log(`    Vendor ID:     ${payload.slice(6, 10)}`);
  console.log(`    Batch Code:    ${payload.slice(10, 12)}`);
  console.log(`    Filament ID:   ${filamentId} (${CREALITY_FILAMENT_NAMES[filamentId] ?? 'Unknown'})`);
  console.log(`    Color (hex):   #${colorHex}`);
  console.log(`    Length Code:   ${lengthCode} (${LENGTH_CODE_TO_WEIGHT[lengthCode] ?? 'Unknown'})`);
  console.log(`    Serial Num:    ${payload.slice(28, 34)}`);
}

// ============================================================================
// Dump printer
// ============================================================================

function printDump(dump) {
  console.log(`  UID:  ${dump.uid ?? 'N/A'}`);
  console.log(`  Type: ${dump.type ?? 'N/A'}`);

  if (dump.pages?.length) {
    console.log(`  Pages (${dump.pages.length} total):`);
    dump.pages.slice(0, 8).forEach((page, i) => {
      console.log(`    Page ${String(i).padStart(3)}: ${page}`);
    });
    if (dump.pages.length > 8) {
      console.log(`    ... and ${dump.pages.length - 8} more pages`);
    }
  }

  if (dump.blocks && Object.keys(dump.blocks).length > 0) {
    const sortedKeys = Object.keys(dump.blocks).sort((a, b) => parseInt(a) - parseInt(b));
    console.log(`  Blocks (${sortedKeys.length} readable):`);
    sortedKeys.slice(0, 8).forEach((k) => {
      console.log(`    Block ${String(k).padStart(3)}: ${dump.blocks[k]}`);
    });
    if (sortedKeys.length > 8) {
      console.log(`    ... and ${sortedKeys.length - 8} more blocks`);
    }
    if (dump.failedBlocks?.length) {
      console.log(`  Failed blocks (unknown keys): ${dump.failedBlocks.join(', ')}`);
    }

    // Try Creality decode
    if (dump.blocks['4'] || dump.blocks['5'] || dump.blocks['6']) {
      console.log('\n  Attempting Creality CFS decode...');
      tryDecodeCreality(dump.blocks);
    }
  }
}

// ============================================================================
// Test 1: HTTP dump
// ============================================================================

async function testHttpDump(readerIndex = 0) {
  console.log('\n[1] HTTP GET /v1/readers/0/dump');
  console.log('-'.repeat(50));

  try {
    const resp = await fetch(`${BASE_URL}/readers/${readerIndex}/dump`);
    if (!resp.ok) {
      const body = await resp.text();
      console.log(`  HTTP dump failed: ${resp.status} ${body.slice(0, 200)}`);
      return false;
    }
    const dump = await resp.json();
    console.log('  HTTP dump response received:');
    printDump(dump);
    return true;
  } catch (err) {
    console.log(`  ERROR: ${err.message}`);
    return false;
  }
}

// ============================================================================
// Test 2: WS dumpCard
// ============================================================================

async function testWsDump(readerIndex = 0) {
  console.log('\n[2] WebSocket dumpCard() command');
  console.log('-'.repeat(50));

  const ws = new NFCAgentWebSocket({ autoReconnect: false });
  try {
    await ws.connect();
    console.log('  Connected to NFC Agent WS');

    const dump = await Promise.race([
      ws.dumpCard(readerIndex),
      new Promise((_, reject) => setTimeout(() => reject(new Error('Timeout')), 15000)),
    ]);

    console.log('  dumpCard() response received:');
    printDump(dump);
    return true;
  } catch (err) {
    console.log(`  dumpCard failed: ${err.message}`);
    return false;
  } finally {
    ws.disconnect();
  }
}

// ============================================================================
// Test 3: WS subscribe with includeRaw
// ============================================================================

async function testWsSubscribeRaw(readerIndex = 0) {
  console.log('\n[3] WebSocket subscribe({ includeRaw: true })');
  console.log('-'.repeat(50));
  console.log('  Tap a card on the reader to see card_detected + card_data events...');
  console.log('  (waiting up to 30 seconds, Ctrl+C to skip)');

  const ws = new NFCAgentWebSocket({ autoReconnect: false });

  return new Promise(async (resolve) => {
    let timer;

    ws.on('card_detected', (event) => {
      console.log(`\n  >> card_detected: UID=${event.card.uid}, Type=${event.card.type}`);
      console.log(`     (waiting for card_data event...)`);
    });

    ws.on('card_data', (event) => {
      console.log(`\n  >> card_data received!`);
      printDump(event);
      clearTimeout(timer);
      ws.disconnect();
      console.log('\n  Subscribe test PASSED');
      resolve(true);
    });

    ws.on('error', (err) => {
      console.log(`  WS error: ${err.message}`);
    });

    try {
      await ws.connect();
      console.log('  Connected to NFC Agent WS');
      await ws.subscribe(readerIndex, { includeRaw: true });
      console.log(`  Subscribed to reader ${readerIndex} with includeRaw: true`);

      timer = setTimeout(() => {
        console.log('\n  No card tapped within 30s — subscribe test skipped');
        ws.disconnect();
        resolve(false);
      }, 30000);
    } catch (err) {
      console.log(`  Connection failed: ${err.message}`);
      resolve(false);
    }
  });
}

// ============================================================================
// Main
// ============================================================================

const args = process.argv.slice(2);
const testArg = args.find((a) => a.startsWith('--test='))?.split('=')[1]
  ?? (args[args.indexOf('--test') + 1]);
const readerIndex = parseInt(
  args.find((a) => a.startsWith('--reader='))?.split('=')[1]
  ?? args[args.indexOf('--reader') + 1]
  ?? '0'
);

const runHttp = !testArg || testArg === 'http' || testArg === 'all';
const runWs = !testArg || testArg === 'ws' || testArg === 'all';
const runSubscribe = testArg === 'subscribe';

console.log('='.repeat(60));
console.log('  NFC Agent Raw Dump Test (JavaScript SDK)');
console.log('='.repeat(60));

if (runHttp) await testHttpDump(readerIndex);
if (runWs) await testWsDump(readerIndex);
if (runSubscribe) await testWsSubscribeRaw(readerIndex);

console.log('\n' + '='.repeat(60));
console.log('  Done');
console.log('='.repeat(60));

File: scripts/test_dump.py
#!/usr/bin/env python3
"""
Test script for raw card dump via Python SDK.

Tests:
  1. HTTP: GET /v1/readers/0/dump
  2. WS: dump_card command
  3. WS: subscribe with include_raw=True (tap a card to trigger)
  4. Creality CFS decode from MIFARE Classic raw blocks

Usage:
    cd /path/to/nfc-agent
    pip install -e sdk/python pycryptodome requests
    python3 scripts/test_dump.py
    python3 scripts/test_dump.py --test subscribe   # also tests include_raw subscribe
"""

from __future__ import annotations

import argparse
import asyncio
import sys

try:
    import requests
except ImportError:
    print("ERROR: pip install requests")
    sys.exit(1)

try:
    from nfc_agent import NFCWebSocket, CardDataEvent, CardRawDump
except ImportError:
    print("ERROR: pip install -e sdk/python")
    sys.exit(1)

BASE_URL = "http://127.0.0.1:32145/v1"

# Creality CFS AES keys (from CrealityMaterialStandard.php)
CREALITY_AES_KEY_DATA = b"H@CFkRnz@KAtBJp2"

CREALITY_FILAMENT_NAMES: dict[str, str] = {
    "010101": "PLA",
    "010201": "PETG",
    "010301": "ABS",
    "010401": "TPU",
    "010501": "ASA",
    "010601": "PC",
    "010701": "PA",
    "010801": "PLA-CF",
    "010901": "PETG-CF",
    "010A01": "ABS-CF",
    "010B01": "PA-CF",
    "010C01": "PLA-Silk",
    "010D01": "PLA-Matte",
}

LENGTH_CODE_TO_WEIGHT: dict[str, str] = {
    "0330": "1000g",
    "0247": "750g",
    "0198": "600g",
    "0165": "500g",
    "0082": "250g",
}


def http_dump(reader_index: int = 0) -> dict | None:
    """Call GET /v1/readers/{n}/dump via HTTP."""
    try:
        resp = requests.get(f"{BASE_URL}/readers/{reader_index}/dump", timeout=15)
        if resp.status_code == 200:
            return resp.json()
        print(f"  HTTP dump failed: {resp.status_code} {resp.text[:200]}")
        return None
    except requests.exceptions.ConnectionError:
        print("  ERROR: Cannot connect to NFC Agent")
        return None


def try_decrypt_creality(blocks: dict[str, str]) -> dict | None:
    """Try to decrypt Creality CFS data from MIFARE Classic blocks 4, 5, 6."""
    try:
        from Crypto.Cipher import AES
    except ImportError:
        print("  (pycryptodome not installed, skipping Creality decode)")
        return None

    encrypted_blocks = []
    for block_num in [4, 5, 6]:
        key = str(block_num)
        if key not in blocks:
            print(f"  Block {block_num} not available — Creality decode skipped")
            return None
        encrypted_blocks.append(blocks[key])

    decrypted = ""
    for i, enc_hex in enumerate(encrypted_blocks):
        enc_bytes = bytes.fromhex(enc_hex)
        cipher = AES.new(CREALITY_AES_KEY_DATA, AES.MODE_ECB)
        dec = cipher.decrypt(enc_bytes).decode("ascii", errors="replace")
        print(f"  Block {4+i} decrypted: {repr(dec)}")
        decrypted += dec

    print(f"\n  Full payload (48 chars): {repr(decrypted)}")

    if len(decrypted) < 34:
        print(f"  Payload too short: {len(decrypted)} chars")
        return None

    payload = decrypted.ljust(48, "0")[:48]
    parsed = {
        "date_code": payload[0:6],
        "vendor_id": payload[6:10],
        "batch_code": payload[10:12],
        "filament_id": payload[12:18].upper(),
        "color_hex": payload[18:24].upper(),
        "length_code": payload[24:28],
        "serial_num": payload[28:34],
        "padding": payload[34:48],
    }

    filament_name = CREALITY_FILAMENT_NAMES.get(parsed["filament_id"], "Unknown")
    weight = LENGTH_CODE_TO_WEIGHT.get(parsed["length_code"], "Unknown")

    print("\n  Parsed Creality fields:")
    print(f"    Date Code:     {parsed['date_code']}")
    print(f"    Vendor ID:     {parsed['vendor_id']}")
    print(f"    Batch Code:    {parsed['batch_code']}")
    print(f"    Filament ID:   {parsed['filament_id']} ({filament_name})")
    print(f"    Color (hex):   #{parsed['color_hex']}")
    print(f"    Length Code:   {parsed['length_code']} ({weight})")
    print(f"    Serial Num:    {parsed['serial_num']}")

    return parsed


def print_dump(dump: dict) -> None:
    """Pretty-print a CardRawDump dict."""
    print(f"  UID:  {dump.get('uid', 'N/A')}")
    print(f"  Type: {dump.get('type', 'N/A')}")

    if pages := dump.get("pages"):
        print(f"  Pages ({len(pages)} total):")
        for i, page in enumerate(pages[:8]):
            print(f"    Page {i:3d}: {page}")
        if len(pages) > 8:
            print(f"    ... and {len(pages) - 8} more pages")

    if blocks := dump.get("blocks"):
        print(f"  Blocks ({len(blocks)} readable):")
        for block_num in sorted(blocks.keys(), key=int)[:8]:
            print(f"    Block {int(block_num):3d}: {blocks[block_num]}")
        if len(blocks) > 8:
            print(f"    ... and {len(blocks) - 8} more blocks")

        failed = dump.get("failedBlocks") or []
        if failed:
            print(f"  Failed blocks (unknown keys): {failed}")

        # Try Creality decode if this looks like a MIFARE Classic card
        if "4" in blocks or "5" in blocks or "6" in blocks:
            print("\n  Attempting Creality CFS decode...")
            try_decrypt_creality(blocks)


def print_card_data_event(event: CardDataEvent) -> None:
    """Pretty-print a CardDataEvent."""
    d = {
        "uid": event.uid,
        "type": event.type,
        "pages": event.pages,
        "blocks": event.blocks,
        "failedBlocks": event.failed_blocks,
    }
    # Remove None values
    d = {k: v for k, v in d.items() if v is not None}
    print_dump(d)


async def test_ws_dump(reader_index: int = 0) -> bool:
    """Test dump_card WS command."""
    print("\n[2] WebSocket dump_card command")
    print("-" * 50)

    try:
        async with NFCWebSocket() as ws:
            print("  Connected to NFC Agent WS")
            try:
                dump = await asyncio.wait_for(ws.dump_card(reader_index), timeout=15)
                print("  dump_card response received:")
                print_dump({
                    "uid": dump.uid,
                    "type": dump.type,
                    "pages": dump.pages,
                    "blocks": dump.blocks,
                    "failedBlocks": dump.failed_blocks,
                })
                return True
            except Exception as e:
                print(f"  dump_card failed: {e}")
                return False
    except Exception as e:
        print(f"  WS connection failed: {e}")
        return False


async def test_ws_subscribe_raw(reader_index: int = 0) -> None:
    """Test subscribe with include_raw=True."""
    print("\n[3] WebSocket subscribe with include_raw=True")
    print("-" * 50)
    print("  Tap a card on the reader to see card_detected + card_data events...")
    print("  (waiting up to 30 seconds, Ctrl+C to skip)")

    received_data = asyncio.Event()

    try:
        async with NFCWebSocket() as ws:
            @ws.on_card_detected
            def handle_detected(event):
                print(f"\n  >> card_detected: UID={event.card.uid}, Type={event.card.type}")
                print(f"     (waiting for card_data event...)")

            @ws.on_card_data
            def handle_data(event: CardDataEvent):
                print(f"\n  >> card_data received!")
                print_card_data_event(event)
                received_data.set()

            await ws.subscribe(reader_index, include_raw=True)
            print(f"  Subscribed to reader {reader_index} with include_raw=True")

            try:
                await asyncio.wait_for(received_data.wait(), timeout=30)
                print("\n  Subscribe test PASSED")
            except asyncio.TimeoutError:
                print("\n  No card tapped within 30s — subscribe test skipped")
    except Exception as e:
        print(f"  WS connection failed: {e}")


async def main() -> int:
    parser = argparse.ArgumentParser(description="Test NFC Agent raw card dump")
    parser.add_argument("--reader", type=int, default=0, help="Reader index (default: 0)")
    parser.add_argument(
        "--test",
        choices=["http", "ws", "subscribe", "all"],
        default="all",
        help="Which test to run (default: all, skips subscribe interaction)",
    )
    args = parser.parse_args()

    print("=" * 60)
    print("  NFC Agent Raw Dump Test (Python SDK)")
    print("=" * 60)

    run_http = args.test in ("http", "all")
    run_ws = args.test in ("ws", "all")
    run_subscribe = args.test == "subscribe"

    # -- Test 1: HTTP dump --
    if run_http:
        print("\n[1] HTTP GET /v1/readers/0/dump")
        print("-" * 50)
        dump = http_dump(args.reader)
        if dump:
            print("  HTTP dump response received:")
            print_dump(dump)
        else:
            print("  No card? Place a card on the reader and try again.")

    # -- Test 2: WS dump_card --
    if run_ws:
        await test_ws_dump(args.reader)

    # -- Test 3: WS subscribe with include_raw --
    if run_subscribe:
        await test_ws_subscribe_raw(args.reader)

    print("\n" + "=" * 60)
    print("  Done")
    print("=" * 60)
    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))
